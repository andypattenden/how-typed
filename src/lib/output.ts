import chalk from 'chalk'
import { formattingColours, subHeadingColours } from '../constants.js'
import type { Counts } from '../@types/index.js'
import { formatPercentageString, formatRating } from './formatting.js'
import { buildTableRow, buildTable } from './table.js'
import { getTotalsCountForFiletype } from './utils.js'
import { getRating } from './ratings.js'

const subheadingUnderlineCharacter = '\u2015'

export function outputHeader(title: string) {
  const bannerGraphicTopRow = `${chalk.bold.bgHex(formattingColours.tsLogo)('      ')}    ${chalk.bold.bgHex(formattingColours.jsLogo)('      ')}`
  const bannerGraphicBottomRow = `${chalk.bold.bgHex(formattingColours.tsLogo).hex(formattingColours.white)('   TS ')} vs ${chalk.bold.bgHex(formattingColours.jsLogo).hex(formattingColours.black)('   JS ')}`

  const headerBoxPadding = 4

  const logoLength = 16
  const rowLength = Math.max(title.length, logoLength) + headerBoxPadding * 2

  const logoSpacing = ' '.repeat(
    Math.max(
      (title.length - logoLength) / 2 + headerBoxPadding,
      headerBoxPadding,
    ),
  )

  // Add extra space to the logo if the title is longer than the logo and odd length
  const addExtraLogoSpace =
    title.length % 2 !== 0 && title.length > logoLength ? ' ' : ''
  // Add extra space to the title if the title is shorter than the logo and odd length
  const addExtraTitleSpace =
    title.length % 2 !== 0 && title.length < logoLength ? ' ' : ''
  const titleSpacing = ' '.repeat(
    Math.max((logoLength - title.length) / 2 + 4, headerBoxPadding),
  )
  const topRow = `\n\u2554${'\u2550'.repeat(rowLength)}\u2557\n`
  const logoRow = `\u2551${logoSpacing}${bannerGraphicTopRow}${logoSpacing}${addExtraLogoSpace}\u2551\n\u2551${logoSpacing}${bannerGraphicBottomRow}${logoSpacing}${addExtraLogoSpace}\u2551\n`

  const titleRow = `\u2551${titleSpacing}${chalk.bold(title)}${titleSpacing}${addExtraTitleSpace}\u2551\n`

  const blankRow = `\u2551${' '.repeat(rowLength)}\u2551\n`
  const bottomRow = `\u255A${'\u2550'.repeat(rowLength)}\u255D\n`

  console.log(`${topRow}${logoRow}${blankRow}${titleRow}${bottomRow}`)
}

export function outputSubheader(title: string) {
  console.log(
    chalk.bold[subHeadingColours.fg](
      `\n${title}\n${subheadingUnderlineCharacter.repeat(title.length)}`,
    ),
  )
}

export function outputReportSummary(counts: Counts) {
  outputSubheader('Summary')

  const overallPercentages = counts.percentages

  const summary = overallPercentages
    ? `Overall, ${chalk.bold(formatPercentageString(overallPercentages.files.ts))} of files and ${chalk.bold(formatPercentageString(overallPercentages.loc.ts))} of lines of code (LoC) are written in TypeScript, earning a ${chalk.bold(formatRating(getRating(overallPercentages.files.ts)))} rating for file coverage and a ${formatRating(getRating(overallPercentages.loc.ts))} rating for LoC coverage.`
    : 'Summary not available'

  // TODO: calculate the best category for files
  // const bestCategoryForLoC = getBestCategoryForLoc(counts)
  // const bestCategoryForFiles = getBestCategoryForFiles(counts)

  // TODO: calculate the categories for improvements

  console.log(summary)
}

export function outputReport(title: string, counts: Counts, json = false) {
  if (counts.total.files === 0) {
    console.log(
      chalk.red('ðŸ™… No JavaScript or TypeScript files found in the directory'),
    )
  } else {
    if (json) {
      console.log(JSON.stringify(counts, null, 2))
      return
    }

    outputToConsole(title, counts)
  }
}

const outputToConsole = (title: string, counts: Counts) => {
  outputHeader(title)

  const tableRows = []

  Object.keys(counts.filetypes)
    .sort()
    .forEach((key) => {
      const count = counts.filetypes[key]
      if (count.total.files > 0) {
        tableRows.push(
          ...buildTableRow(
            count.label,
            count.js,
            count.ts,
            count.percentages,
            count.total,
            count.score,
          ),
        )
      }
    })

  const totalJsCounts = getTotalsCountForFiletype(counts, 'js')
  const totalTsCounts = getTotalsCountForFiletype(counts, 'ts')

  tableRows.push(
    ...buildTableRow(
      'OVERALL',
      totalJsCounts,
      totalTsCounts,
      counts.percentages,
      counts.total,
      counts.score,
      true,
    ),
  )

  const table = buildTable(tableRows)

  console.log(table.toString())

  outputReportSummary(counts)
}
