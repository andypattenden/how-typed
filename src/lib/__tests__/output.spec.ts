import type { MockInstance } from 'vitest'
import * as output from '../output.js'

vi.stubGlobal('console', { log: vi.fn() })

describe('output', () => {
  let log: MockInstance<typeof console.log>

  beforeEach(() => {
    log = vi.mocked(console.log)
    log.mockReset()
  })

  describe('outputHeader', () => {
    it.each([['Title'], ['Another Title'], ['A title with an odd length!']])(
      'should output a header with %s',
      (title) => {
        output.outputHeader(title)

        log.mock.calls.forEach((call) => {
          expect(call).toMatchSnapshot()
        })
      },
    )
  })

  describe('outputSubheader', () => {
    it('should output a subheader', () => {
      output.outputSubheader('Subheader')

      log.mock.calls.forEach((call) => {
        expect(call).toMatchSnapshot()
      })
    })
  })

  describe('outputReportSummary', () => {
    it('should output a report summary', () => {
      output.outputReportSummary({
        percentages: {
          files: {
            ts: 60,
            js: 40,
          },
          loc: {
            ts: 60,
            js: 40,
          },
        },
        total: {
          files: 26,
          loc: 1000,
        },
        score: 59.31,
        filetypes: {
          plain: {
            label: 'Plain JS/TS',
            js: {
              files: 4,
              loc: 6,
            },
            ts: {
              files: 6,
              loc: 8,
            },
            total: {
              files: 10,
              loc: 14,
            },
            percentages: {
              files: {
                ts: 60,
                js: 40,
              },
              loc: {
                ts: 70,
                js: 30,
              },
            },
            score: 67,
          },
          vue: {
            label: 'Vue SFC',
            js: {
              files: 7,
              loc: 12,
            },
            ts: {
              files: 9,
              loc: 16,
            },
            total: {
              files: 16,
              loc: 28,
            },
            percentages: {
              files: {
                ts: 56.25,
                js: 43.75,
              },
              loc: {
                ts: 75,
                js: 25,
              },
            },
            score: 69.38,
          },
        },
      })

      log.mock.calls.forEach((call) => {
        expect(call).toMatchSnapshot()
      })
    })
  })

  describe('outputReport', () => {
    it('should output a report', () => {
      output.outputReport('Title', {
        total: {
          files: 26,
          loc: 1000,
        },
        percentages: {
          files: {
            ts: 57.69,
            js: 42.31,
          },
          loc: {
            ts: 60,
            js: 40,
          },
        },
        score: 59.31,
        filetypes: {
          plain: {
            label: 'Plain JS/TS',
            js: {
              files: 4,
              loc: 6,
            },
            ts: {
              files: 6,
              loc: 8,
            },
            total: {
              files: 10,
              loc: 14,
            },
            percentages: {
              files: {
                ts: 60,
                js: 40,
              },
              loc: {
                ts: 70,
                js: 30,
              },
            },
            score: 67,
          },
          vue: {
            label: 'Vue SFC',
            js: {
              files: 7,
              loc: 12,
            },
            ts: {
              files: 9,
              loc: 16,
            },
            total: {
              files: 16,
              loc: 28,
            },
            percentages: {
              files: {
                ts: 56.25,
                js: 43.75,
              },
              loc: {
                ts: 75,
                js: 25,
              },
            },
            score: 69.38,
          },
        },
      })

      log.mock.calls.forEach((call) => {
        expect(call).toMatchSnapshot()
      })
    })

    it('should output a report with no files found', () => {
      output.outputReport('Title', {
        total: {
          files: 0,
          loc: 0,
        },
        percentages: {
          files: {
            ts: 0,
            js: 0,
          },
          loc: {
            ts: 0,
            js: 0,
          },
        },
        score: 0,
        filetypes: {
          plain: {
            label: 'Plain JS/TS',
            js: {
              files: 0,
              loc: 0,
            },
            ts: {
              files: 0,
              loc: 0,
            },
            total: {
              files: 0,
              loc: 0,
            },
            percentages: {
              files: {
                ts: 0,
                js: 0,
              },
              loc: {
                ts: 0,
                js: 0,
              },
            },
            score: 0,
          },
          vue: {
            label: 'Vue SFC',
            js: {
              files: 0,
              loc: 0,
            },
            ts: {
              files: 0,
              loc: 0,
            },
            total: {
              files: 0,
              loc: 0,
            },
            percentages: {
              files: {
                ts: 0,
                js: 0,
              },
              loc: {
                ts: 0,
                js: 0,
              },
            },
            score: 0,
          },
        },
      })

      log.mock.calls.forEach((call) => {
        expect(call).toMatchSnapshot()
      })
    })
  })
})
