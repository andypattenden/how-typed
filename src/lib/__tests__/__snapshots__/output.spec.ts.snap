// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`output > outputHeader > should output a header with A title with an odd length! 1`] = `
[
  "
╔═══════════════════════════════════╗
║         [1m[48;2;49;120;198m      [49m[22m    [1m[48;2;240;219;79m      [49m[22m          ║
║         [1m[48;2;49;120;198m[38;2;255;255;255m   TS [39m[49m[22m vs [1m[48;2;240;219;79m[38;2;0;0;0m   JS [39m[49m[22m          ║
║                                   ║
║    [1mA title with an odd length![22m    ║
╚═══════════════════════════════════╝
",
]
`;

exports[`output > outputHeader > should output a header with Another Title 1`] = `
[
  "
╔════════════════════════╗
║    [1m[48;2;49;120;198m      [49m[22m    [1m[48;2;240;219;79m      [49m[22m    ║
║    [1m[48;2;49;120;198m[38;2;255;255;255m   TS [39m[49m[22m vs [1m[48;2;240;219;79m[38;2;0;0;0m   JS [39m[49m[22m    ║
║                        ║
║     [1mAnother Title[22m      ║
╚════════════════════════╝
",
]
`;

exports[`output > outputHeader > should output a header with Title 1`] = `
[
  "
╔════════════════════════╗
║    [1m[48;2;49;120;198m      [49m[22m    [1m[48;2;240;219;79m      [49m[22m    ║
║    [1m[48;2;49;120;198m[38;2;255;255;255m   TS [39m[49m[22m vs [1m[48;2;240;219;79m[38;2;0;0;0m   JS [39m[49m[22m    ║
║                        ║
║         [1mTitle[22m          ║
╚════════════════════════╝
",
]
`;

exports[`output > outputReport > should output a report 1`] = `
[
  "
╔════════════════════════╗
║    [1m[48;2;49;120;198m      [49m[22m    [1m[48;2;240;219;79m      [49m[22m    ║
║    [1m[48;2;49;120;198m[38;2;255;255;255m   TS [39m[49m[22m vs [1m[48;2;240;219;79m[38;2;0;0;0m   JS [39m[49m[22m    ║
║                        ║
║         [1mTitle[22m          ║
╚════════════════════════╝
",
]
`;

exports[`output > outputReport > should output a report 2`] = `
[
  "[90m┌─────────────[39m[90m┬────────[39m[90m┬─────[39m[90m┬─────[39m[90m┬───────[39m[90m┬────────[39m[90m┬─────────┐[39m
[90m│[39m[31m [1m[94mFile Type[31m[22m   [39m[90m│[39m[31m [1m[94mMetric[31m[22m [39m[90m│[39m[31m [1m[43m[30m JS[31m[49m[22m [39m[90m│[39m[31m [1m[44m[37m TS[31m[49m[22m [39m[90m│[39m[31m [1m[94mTotal[31m[22m [39m[90m│[39m[31m [1m[44m[37m TS[31m[49m [94m%[31m[22m  [39m[90m│[39m[31m [1m[94mRating[31m[22m  [39m[90m│[39m
[90m├─────────────[39m[90m┼────────[39m[90m┼─────[39m[90m┼─────[39m[90m┼───────[39m[90m┼────────[39m[90m┼─────────┤[39m
[90m│[39m Plain JS/TS [90m│[39m Files  [90m│[39m 4   [90m│[39m 6   [90m│[39m 10    [90m│[39m [1m[48;2;51;51;0m[38;2;255;255;0m60.00%[39m[49m[22m [90m│[39m         [90m│[39m
[90m│[39m             [90m├────────[39m[90m┼─────[39m[90m┼─────[39m[90m┼───────[39m[90m┼────────[39m[90m┤[39m [1m[48;2;51;51;0m[38;2;255;255;0mGood[39m[49m[22m 👍 [90m│[39m
[90m│[39m             [90m│[39m LoC    [90m│[39m 6   [90m│[39m 8   [90m│[39m 14    [90m│[39m [1m[48;2;51;51;0m[38;2;255;255;0m70.00%[39m[49m[22m [90m│[39m         [90m│[39m
[90m├─────────────[39m[90m┼────────[39m[90m┼─────[39m[90m┼─────[39m[90m┼───────[39m[90m┼────────[39m[90m┼─────────┤[39m
[90m│[39m Vue SFC     [90m│[39m Files  [90m│[39m 7   [90m│[39m 9   [90m│[39m 16    [90m│[39m [1m[48;2;51;26;0m[38;2;255;165;0m56.25%[39m[49m[22m [90m│[39m         [90m│[39m
[90m│[39m             [90m├────────[39m[90m┼─────[39m[90m┼─────[39m[90m┼───────[39m[90m┼────────[39m[90m┤[39m [1m[48;2;51;51;0m[38;2;255;255;0mGood[39m[49m[22m 👍 [90m│[39m
[90m│[39m             [90m│[39m LoC    [90m│[39m 12  [90m│[39m 16  [90m│[39m 28    [90m│[39m [1m[48;2;51;77;0m[38;2;173;255;47m75.00%[39m[49m[22m [90m│[39m         [90m│[39m
[90m├─────────────[39m[90m┼────────[39m[90m┼─────[39m[90m┼─────[39m[90m┼───────[39m[90m┼────────[39m[90m┼─────────┤[39m
[90m│[39m [1mOVERALL[22m     [90m│[39m [1mFiles[22m  [90m│[39m [1m11[22m  [90m│[39m [1m15[22m  [90m│[39m [1m26[22m    [90m│[39m [1m[1m[48;2;51;26;0m[38;2;255;165;0m57.69%[39m[49m[22m[1m[22m [90m│[39m         [90m│[39m
[90m│[39m             [90m├────────[39m[90m┼─────[39m[90m┼─────[39m[90m┼───────[39m[90m┼────────[39m[90m┤[39m [1m[1m[48;2;51;26;0m[38;2;255;165;0mFair[39m[49m[22m[1m 🤏[22m [90m│[39m
[90m│[39m             [90m│[39m [1mLoC[22m    [90m│[39m [1m18[22m  [90m│[39m [1m24[22m  [90m│[39m [1m1000[22m  [90m│[39m [1m[1m[48;2;51;51;0m[38;2;255;255;0m60.00%[39m[49m[22m[1m[22m [90m│[39m         [90m│[39m
[90m└─────────────[39m[90m┴────────[39m[90m┴─────[39m[90m┴─────[39m[90m┴───────[39m[90m┴────────[39m[90m┴─────────┘[39m",
]
`;

exports[`output > outputReport > should output a report 3`] = `
[
  "[1m[34m[39m[22m
[1m[34mSummary[39m[22m
[1m[34m―――――――[39m[22m",
]
`;

exports[`output > outputReport > should output a report 4`] = `
[
  "Overall, [1m[1m[48;2;51;26;0m[38;2;255;165;0m57.69%[39m[49m[22m[1m[22m of files and [1m[1m[48;2;51;51;0m[38;2;255;255;0m60.00%[39m[49m[22m[1m[22m of lines of code (LoC) are written in TypeScript, earning a [1m[1m[48;2;51;26;0m[38;2;255;165;0mFair[39m[49m[22m[1m 🤏[22m rating for file coverage and a [1m[48;2;51;51;0m[38;2;255;255;0mGood[39m[49m[22m 👍 rating for LoC coverage.",
]
`;

exports[`output > outputReport > should output a report with no files found 1`] = `
[
  "[31m🙅 No JavaScript or TypeScript files found in the directory[39m",
]
`;

exports[`output > outputReportSummary > should output a report summary 1`] = `
[
  "[1m[34m[39m[22m
[1m[34mSummary[39m[22m
[1m[34m―――――――[39m[22m",
]
`;

exports[`output > outputReportSummary > should output a report summary 2`] = `
[
  "Overall, [1m[1m[48;2;51;51;0m[38;2;255;255;0m60.00%[39m[49m[22m[1m[22m of files and [1m[1m[48;2;51;51;0m[38;2;255;255;0m60.00%[39m[49m[22m[1m[22m of lines of code (LoC) are written in TypeScript, earning a [1m[1m[48;2;51;51;0m[38;2;255;255;0mGood[39m[49m[22m[1m 👍[22m rating for file coverage and a [1m[48;2;51;51;0m[38;2;255;255;0mGood[39m[49m[22m 👍 rating for LoC coverage.",
]
`;

exports[`output > outputSubheader > should output a subheader 1`] = `
[
  "[1m[34m[39m[22m
[1m[34mSubheader[39m[22m
[1m[34m―――――――――[39m[22m",
]
`;
